src/helpers.coffee

```diff
@@ -244,6 +248,53 @@ exports.nameWhitespaceCharacter = (string) ->
     when '\t' then 'tab'
     else string
 
+exports.locationDataToBabylon = ({first_line, first_column, last_line, last_column, range}) ->
+  loc:
+    start:
+      line: first_line + 1
+      column: first_column
+    end:
+      line: last_line + 1
+      column: last_column + 1
+  range: [
+    range[0]
+    range[1] + 1
+  ]
+  start: range[0]
+  end: range[1] + 1
+
+exports.babylonLocationFields = ['loc', 'range', 'start', 'end']
+
+exports.isArray = isArray = (obj) -> Array.isArray obj
+exports.isFunction = (obj) -> Object::toString.call(obj) is '[object Function]'
+exports.isNumber = isNumber = (obj) -> Object::toString.call(obj) is '[object Number]'
+exports.isString = isString = (obj) -> Object::toString.call(obj) is '[object String]'
+exports.isBoolean = isBoolean = (obj) -> obj is yes or obj is no or Object::toString.call(obj) is '[object Boolean]'
+exports.isPlainObject = (obj) -> typeof obj is 'object' and !!obj and not isArray(obj) and not isNumber(obj) and not isString(obj) and not isBoolean(obj)
+
+# Converts a number, string, or node (Value/NumberLiteral/unary +/- Op) to its
+# corresponding number value.
+exports.getNumberValue = (number) ->
+  return number if isNumber number
+  invert = no
+  unless isString number
+    number = do ->
```

Comment:
As far as `do ->` style goes, I’m generally not a fan. I appreciate how it encapsulates logic, like if you had made a `getNumberValue` function and called `number = getNumberValue number`, but to do it with an anonymous function feels like it defeats the purpose, as it’s not really separated from the parent function’s flow. Especially when the logic you’re encapsulating is short, it feels unnecessary. This one could be written as:

```coffee
  unless isString number
    number = number.unwrap()
    if number.operator
      invert = yes if number.operator is '-'
      number = number.first.unwrap().value
    else
      number = number.value
```

Which is actually shorter than the `do ->` version and arguably more readable, as the `do` version relies on two implicit returns which casual readers might not notice if they didn’t notice the `do ->` further up. The `do` keyword is rarely used to begin with, and I’d never seen `do ->` before we started discussing it, so this isn’t a pattern that I think most people are familiar with.

I guess if it were me, I would either keep it inline like my example here if the logic were brief, or split it out into a function and call that function, as the act of naming the function helps to explain what it is we’re encapsulating and why it’s getting modularized. The function could still be declared right here, inside this same parent function; if it’s not used anywhere else in the codebase, there’s no reason to declare it in a helpers file or something like that.

Anyway this is all style, I’m not saying your way is wrong or that it has to be done the way I’m describing. Just personal preference.

One more way to do it would be to take advantage of `if` assignment, that’s closer to what you’re looking for:

```coffee
  unless isString number
    number = number.unwrap()
    number = if number.operator
      invert = yes if number.operator is ‘-’
      number.first.unwrap().value
    else
      number.value
```

Reply:
@GeoffreyBooth all your suggested guidelines make sense, updated to remove `do ->` here